test/puppetlabs/puppetdb/generative/generators.clj

```diff
@@ -0,0 +1,297 @@
+(ns puppetlabs.puppetdb.generative.generators
+  (:require [clojure.test.check.generators :as gen]
+            [com.gfredericks.test.chuck.generators :as chuck-gen]
+            [clj-time.core :as time :refer [now]]))
+
+
+;;; common data types and combinators
+(defn nilable [g]
+  (gen/one-of [(gen/return nil) g]))
+
+(defn string+
+  ([]
+   (gen/fmap clojure.string/join (gen/vector gen/char-alphanumeric 1 255)))
+  ([max-len]
+   (gen/fmap clojure.string/join (gen/vector gen/char-alphanumeric 1 max-len))))
+
+(defn pg-smallint [] (gen/choose -32768 32767))
+(defn pg-integer [] (gen/choose -2147483648 2147483647))
+(defn pg-pos-integer [] (gen/choose 1 2147483647))
+(defn pg-bigint [] (gen/choose -9223372036854775808 9223372036854775807))
+
+(defn uuid-string [] (gen/fmap str gen/uuid))
+
+(defn datetime []
+  (gen/fmap clj-time.coerce/from-long
+            (gen/choose 50000000000 2000000000000)))
+
+
+
+;;; catalog generator
+
+(defn certname [] (string+))
+(defn environment [] (string+))
+(defn transaction_uuid [] (uuid-string))
+(defn catalog_uuid [] (uuid-string))
+(defn producer_timestamp [] (string+))
+
+(defn producer [] (string+))
+(defn code_id [] (nilable (string+)))
+
+(defn version [] (string+))
+
+(defn resource-type [] (string+))
+(defn resource-title [] (string+))
+(defn resource_spec []
+  (gen/hash-map :type (resource-type)
+                :title (resource-title)))
+
+(defn source [] (resource_spec))
+(defn target [] (resource_spec))
+(defn relationship []
+  (gen/one-of (map gen/return
+                   ["contains"
+                    "required-by"
+                    "notifies"
+                    "before"
+                    "subscription-of"])))
+
+(defn edge []
+  (gen/hash-map :source (source)
```

Comment:
I think this one is actually stale; see the resource-tree generator and resource-tree-to-edge-list function for the real thing. It only makes trees, but at least it's sort of real.